README.md
# Runescape Item Investment Analysis

Providing investors with item-by-item investment analysis for Runescape items.
 

# Table Of Contents
1. Introduction
2. Purpose and Summary
3. Architecture Diagram
4. Deployment Diagram
5. Sources of Data
6. Backend Component (Daily ETL)
7. Frontend Component (App)
8. Completed and Future Improvements





## Introduction 

Old School Runescape (OSRS) is an online game released in February 2013 as a reboot of the 2007 version of Runescape. Millions of people play the game, which offers a wide variety of activities like going on quests, fighting monsters and other players, and leveling up your character's skills. OSRS players can trade each other items for gold coins, and the game has an automated trading system so players can buy or sell items without having to meet in the same location to conduct the trade.

The automated trading system is called the Grand Exchange. One player puts their item into the system, one player their coins, and it conducts the trade. Because almost all trades happen on the Grand Exchange, players often in try to observe trends in price, and invest in items that will go up. Some players make millions of coins each day by choosing the correct items. 

How do they choose which items to invest in?



## Purpose and Summary

The purpose of this project is to help an investor decide if a specific item is a good investment or not. This project aims to be an investment analysis tool that gives investment analysis on an item-by-item basis. It does so by running a daily market-wide analysis on a backend server (ETL), then sending its results to the frontend server (App) where it is accessible by users. Users search for a specific item and the daily market-wide analysis is applied to make an item-specific prediction. The item's most current data is also presented alongside a visualization of historical data. Below are the steps that this project takes.

1. Two weeks' worth of price and volume data is scraped daily with Scrapy. Pandas is used to analyze 16 attributes for each item.
``` 
If the item has changed in (price, volume)
    consecutively in the past (2, 3, 5, 7) days
How much the item has changed in (price, volume)
    compared to (1, 3, 7, 14) days ago
```

2. A data summary is created by aggregating the results of data analysis.
3. Airflow sends the data to the App Server that the user interacts with.
4. When a user searches for an item, the today's data summary is applied to that item's attributes to predict how much it will change in price over the next day.
5. Item is searched and data is extracted from ge-tracker.com to provide user with the most up-to-date information on that item.
6. A graph of price and volume over the past two weeks is presented. Price and volume are plotted on the same graph to show possible interactions.

Example outputs are [shown below](https://github.com/KenanBiren/OSRS_Item_Investment_Project#serve-analysis-data)

## Architecture Diagram


<img width="700" alt="Screen Shot 2022-11-06 at 10 05 40 PM" src="https://user-images.githubusercontent.com/116853630/200237471-012691ed-5d99-453a-b551-9af21dc6e4da.png">


## Deployment Diagram


<img width="700" alt="Screen Shot 2022-11-06 at 10 09 30 PM" src="https://user-images.githubusercontent.com/116853630/200394038-4bb37d28-cb3c-4227-a589-6ddcfd7396da.png">



## Sources of Data
There are many sources that give data on items that OSRS investors might want. Here are the sources of data used in this project (showing example item Abyssal whip).


[Official Old School Runescape Website](https://secure.runescape.com/m=itemdb_oldschool/Abyssal+whip/viewitem?obj=4151)

This website shows the daily average price and volume per item. This source gives daily average values for price and volume data, unlike the next two sources which give most-up-to-date values. Therefore this website is used as the data source for the daily-running investment analysis, as the analysis is calculated over two weeks' of data.


[Old School Wiki Database](https://prices.runescape.wiki/osrs/item/4151)

This site shows up-to-date price/volume data and a few other basic fields like "Buy Limit". This site is a fast API that is used to get a master list of all items in the game. Users of this site have the ability to use the API to pre-filter based on specific values (this may be utiilized in the future).


[ge-tracker.com](https://www.ge-tracker.com/item/abyssal-whip)

This site shows near-real time data. This site is used because it includes other data fields such as "Buying Quantity (1 hour)" and "Selling Quantity (1 hour)" which could be very useful information for investors.


## Backend Component (Daily ETL)
    Extract: Daily Scrape
    Transform: Daily Analysis
    Load: Upload to App Server


#### Mention: The backend of this project is based off of my [previous ELT pipeline](https://github.com/Kenan-Biren/OSRS_Investment_Project)


A major goal of that project was to develop more experience with using cloud technologies (AWS) in a data pipeline. Upon finishing that project, I wanted to implement a CI/CD workflow Github and Docker, but I didn't really know what improvements I wanted to implement.
After some thought I realized that I could make this a lot more interesting by providing users an item-by-item analysis instead of a general daily recommendations list. Having a more defined focus for the end product has allowed me to easily think ahead to what features users might want, and decide how to restructure the project to leverage the use of Github and Docker in a CI/CD workflow. 

My next steps are to improve the accuracy of the daily investment analysis (planning to use factorial analysis) and adding a user interface (likely website first, thinking about an app). 




### Extract: Daily Scrape

This project uses the Scrapy framework to extract data from web sources. There are two spiders, one scrapes one day's worth of data and the other scrapes two weeks' worth of data. The two week spider is for the very first scrape or if the pipeline needs to be restarted. A post-scrape script goes along with each spider, which takes the spider output and formats them into 14day_price.csv and 14day_vol.csv to prepare for analysis. 14day_price.csv and 14day_vol.csv contain the past 14 days of price and volume data for all items, with columns names (name, *TODAY'S DATE*, *YESTERDAY'S DATE*,...)



### Transform: Daily Analysis

The daily analysis takes the past 14 days of raw price and volume data and calculates attributes for each item.


Calculated Attributes:

    two_day_run_p = whether daily trading price has increased consecutively in the past two days. 1 = True, 0 = False
    two_day_run_v = whether daily volume has increased consecutively in the past two days. 1 = True, 0 = False
    three_day_run_p = ...                                                   past three days 
    three_day_run_v = ...
    five_day_run_p = ...
    five_day_run_v = ...
    seven_day_run_p = ...
    seven_day_run_v = ...

    one_day_avg_p = % change in daily average price compared to yesterday
    one_day_avg_v = % change in volume compared to yesterday
    three_day_avg_p = ...                            compared to three days ago
    three_day_avg_v = ...
    seven_day_avg_p = ...
    seven_day_avg_v = ...
    fourteen_day_avg_p = ...
    fourteen_day_avg_v = ...


The average effect of each attribute on price change is calculated by averaging recent price change (same as one_day_avg_p) across all items where that factor applies. This produces a data summary which lists each attribute, and how much % change it is expected to make in items that have that attribute. For example, if the average price change for all items with two_day_run_p=1 (True) is 2%, a specific item that has two_day_run_p=1 will be assumed to increase 2% in the next day.

Data Summary:


|date|two_day_run_p|three_day_run_p|five_day_run_p|seven_day_run_p|two_day_run_v|three_day_run_v|five_day_run_v|seven_day_run_v|one_day_avg_p|three_day_avg_p|seven_day_avg_p|fourteen_day_avg_p|one_day_avg_v|three_day_avg_v|seven_day_avg_v|fourteen_day_avg_v|
|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|
|2022/10/22|0.00453|0.00195|0.00105|0.00037|-0.00091|0|0.0|0.0|0.00198|0.00198|0.00093|0.00154|-0.00116|0.00011|0.00069|-0.00057|



Later when a user searches for an item, that item's attribute data is multiplied by the average effect of every factor. This results in a prediction of % price increase (or decrease).


### Load: Upload to App Server
Airflow transfers today's investment analysis to the App Server via the [SFTPOperator](https://airflow.apache.org/docs/apache-airflow-providers-sftp/stable/_api/airflow/providers/sftp/operators/sftp/index.html)





## Frontend Component (App)
    Read User Input
    Serve Analysis Data
    Serve Near-Real Data

### Read User Input 

User is prompted to search for an item. If no item matches, the user given a list of a couple item names that they might have been trying to spell. Right after a match is made, the local filesystem is checked to see when data has been updated, printing a warning for the user if it is out of date.

### Serve Analysis Data 

Pandas library is used to multiply item name's analysis table data with today's data summary, giving a prediction for how much it will go up in price within the next day. A table is printed explaining some of the logic behind the calculation. A graph is created comparing that item's price and volume data for the past two weeks.

| Attribute  | Has Effect? | Effect
| ------------- | ------------- | ------------- |
| two_day_run_p  | Yes  | 0.00453 |
| three_day_run_p  | Yes  | 0.0195 |
| five_day_run_p  | Yes  | 0.00105 |
| seven_day_run_p  | No  | 0 |
| two_day_run_v  | No  | 0 |
| three_day_run_v  | No  | 0 |
| five_day_run_v  | No  | 0 |
| seven_day_run_v  | No  | 0 |
| one_day_avg_p  | Yes  | 0.00198 |
| three_day_avg_p  | Yes  | -0.00198 |
| seven_day_avg_p  | Yes  | -0.00093 |
| fourteen_day_avg_p  | Yes  | -0.00154 |
| one_day_avg_v  | Yes  | 0.00116 |
| three_day_avg_v  | Yes  | -0.00011 |
| seven_day_avg_v  | Yes  | 0.00069 |
| fourteen_day_avg_v  | Yes  | 0.00057 |
| Total |  | 2.492% |

<img width="638" alt="price_volume_graph" src="https://user-images.githubusercontent.com/116853630/198848946-8671cc9a-c59a-4a06-914d-1e3a10bb4b7f.png">


### Serve Near-real Data: 
After analysis data is served, a Python script using the Requests library scrapes near-real-time data from [ge-tracker.com](https://www.ge-tracker.com/item/abyssal-whip)


    Current Price: 1,554,956 (average price item being bought/sold at)

    Current Sell Price: 1,567,979 (price sellers are placing their item)

    Current Offer Price: 1,545,259 (price buyers are offering for item)

    Selling Quantity (1 hour): 74 (amount being placed up for sale)

    Buying Quantity (1 hour): 84 (amount being sold)

    Buy/Sell Ratio: 1.14 (bought per hour/sold per hour)

    Tax: -15,679 (tax on seller)

    Buy Limit: 70 (limit within 4 hours)

   
#### Note: There is a difference between buy and sell prices because the Grand Exchange works as a double auction. 


## Completed and Future Improvements
~~v1.0.0~~ : Completed first project version with functional scripts (no Docker or Airflow)

~~v2.0.0~~ : Dockerized project and implemented CI/CD workflow
```
Restructured project repository for ease of implementing Docker

Rewrote scripts using the Selenium package to use Requests instead (Selenium + Docker is a headache)

Added Dockerfiles for project components in /build/ folder

Implemented Github workflows to automatically build, verify, and send Docker images to Docker Hub (from Github "test" and "main" branches)

Set up Airflow DAGs to automatically delete old Docker resources and pull the appropriate images from Docker Hub before runtime

Eliminated the use of Amazon Simple Storage Service (S3) to improve runtime efficiency and make it easier for others to run this project

        - Deleted scripts / functions that interacted with S3
        - Docker images utilize bind mounts, backend-frontend file transfer is now handled in one operation with Airflow STFPOperator
        - Replaced any lost backend functionalities with new scripts (set_scrape_type.py, trigger_scrape.py)
        
Added basic runtime test scripts for backend (post_scrape_test.py, analysis_test.py)
```


v2.1.0 : Diagram possible failure modes and expand current runtime tests to ensure 100% coverage

v2.2.0 : Implement factorial design of experiments in daily data analysis

v3.0.0 : Implement a user interface


CHANGELOG.txt
FORMAT:

New Version Number: vX.X.X
Date, Time: YYYY-MM-DD HHmm timezone
Author: name
Title: (MAJOR/MINOR/BUGFIX) title
Description: detailed description of changes
--------------------------------------------------------------------------------



UPCOMING CHANGES:
    -ensure 100% test coverage
    -implement factorial design during daily analysis
    -create a user interface


New Version: v2.0.0
Date, Time: 2022-11-07 1223 PST
Author: Kenan Biren
Title: Dockerized project and implemented CI/CD workflow
Description: 

Restructured project repository for ease of implementing Docker

Rewrote scripts using the Selenium package to use Requests instead (Selenium + Docker is a headache)

Added Dockerfiles for project components in /build/ folder

Implemented Github workflows to automatically build, verify, and send Docker images to Docker Hub (from Github "test" and "main" branches)

Set up Airflow DAGs to automatically delete old Docker resources and pull the appropriate images from Docker Hub before runtime

Eliminated the use of Amazon Simple Storage Service (S3) to improve runtime efficiency and make it easier for others to run this project

        - Deleted scripts / functions that interacted with S3
        - Docker images utilize bind mounts, backend-frontend file transfer is now handled in one operation with Airflow STFPOperator
        - Replaced any lost backend functionalities with new scripts (set_scrape_type.py, trigger_scrape.py)
        
Added basic runtime test scripts for backend (post_scrape_test.py, analysis_test.py)



------------------------------------------------------------------------------------------------------------------------------------------------


New Version: v1.0.0
Date, Time: 2022-10-31 1822 PST
Author: Kenan Biren
Title: Implemented first version with functioning scripts (no Docker or Airflow)
Description: Uploaded all (backend and frontend) functional scripts, without any scheduling or CI/CD implementation. Backend: extraction spidersusing Scrapy, analysis scripts using Pandas. Frontend: App script to serve user.
--------------------------------------------------------------------------------






ex.


New Version: v1.1.1
Date, Time: 2022-06-15 1743 PST
Author: Kenan Biren
Title: (BUGFIX) Bug on post_scrape_analysis.py
Description: ...
--------------------------------------------------------------------------------


New Version: v1.1.0
Date, Time: 2022-06-01 0455 PST
Author: Kenan Biren
Title: (MINOR) Improve run time efficiency of process_backend.py
Description: ...
--------------------------------------------------------------------------------


New Version: v1.0.0
Date, Time: 2022-05-26 1351 PST
Author: Kenan Biren
Title: (MAJOR) Implement user interface
Description: ...
--------------------------------------------------------------------------------


New Version: v0.0.0
...


src/frontend/app/app.py
import matplotlib
matplotlib.use("AGG")
import matplotlib.pyplot as plt
import csv
import time
import datetime
import timedelta
import os
import pandas as pd
from tabulate import tabulate
import requests


# prompt user for an item name. If no name matches, give some suggestions
# of items they might have meant to search for
def read_input():
    item_list = []
    print(os.getcwd())
    print(os.listdir())
                        # get current item masterlist
    with open('data/14day_price.csv', mode='r') as f:
        reader = csv.reader(f)
        for row in reader:
            item_list.append(row[0])

                # prompt input
    while(True):
        item = input('What item would you like to search?\n')
        item = item.lower()
        item = item.capitalize()
        print('item: ' + item)
        t = 0
        if item in item_list:
            print('item found! item: ')
            return item
            break
        else:                       # invalid input -> suggestion
            print('Invalid input, please try again. Did you mean:')
            item_split = item.split()
            longest_word = max(item_split, key=len)     # check for item names
            for i in range(len(item_list)):             # containing longest
                if longest_word in item_list[i]:        # word of user's invalid    
                    print(item_list[i])                 # input
                    t = t + 1
                if t == 5:
                    break


item_name = read_input()

        # check if local data folder has been updated with today's data
def check_files():
    current_date = datetime.date.today()
    cd_str = current_date.strftime('%Y/%m/%d')
    yest_date = current_date - datetime.timedelta(days=1)
    yest_str = yest_date.strftime('%Y/%m/%d')

    

    filePath = "data/"
    last_updated = os.path.getmtime(filePath)
            # convert seconds since epoch to readable timestamp
    last_updated_str = time.strftime('%Y/%m/%d', time.localtime(last_updated))
   

            # warn user if data has not been updated recently
    if last_updated_str != cd_str:
        if last_updated_str == yest_str:
            print("Today's daily investment analysis has not yet been updated. Using yesterday's.\n")
        else:
            print("Daily investment analysis has not been updated. Using old data.\n")








def analysis_data(item_name):
    working_df = {}
            # read data folder files into dataframes
    price_df = pd.read_csv('data/14day_price.csv')
    vol_df = pd.read_csv('data/14day_vol.csv')
    analysis_df = pd.read_csv('data/analysis_table.csv')
    summary_df = pd.read_csv('data/data_summary.csv')

    print(summary_df)

    #------------------------------------------------------------
            # print a two-y-axis graph showing price and volume data
    data = []
    x_list = []
            # get two-weeks' price and volume data as lists
    price_row_df = price_df.loc[price_df['name'] == item_name]
    vol_row_df = vol_df.loc[vol_df['name'] == item_name]

    for num in range(15):
        x_list.append(num)

   
            # get rid of item name in price list
    yp_list = price_row_df.loc[price_row_df['name'] == item_name].values.tolist()
    new_yp_list = yp_list[0]
    new_yp_list.reverse()
    new_yp_list.pop()
    new_yp_list.reverse()
   
            # get rid of item name in volume list
    yv_list = vol_row_df.loc[vol_row_df['name'] == item_name].values.tolist()
    new_yv_list = yv_list[0]
    new_yv_list.reverse()
    new_yv_list.pop()
    new_yv_list.reverse()
   

    for n in range(len(x_list)):
        data.append([x_list[n], new_yp_list[n], new_yv_list[n]])


    graph_df = pd.DataFrame(data, columns=['days_ago', 'price', 'volume'])



    fig,ax = plt.subplots()
    # make a plot
    ax.plot(graph_df.days_ago,
            graph_df.price,
            color="red",
            marker="o")
    # set x-axis label
    ax.set_xlabel("days ago", fontsize = 14)
    # set y-axis label
    ax.set_ylabel("price",
                color="red",
                fontsize=14)

    ax2=ax.twinx()
    # make a plot with different y-axis using second axis object
    ax2.plot(graph_df.days_ago, graph_df["volume"],color="blue",marker="o")
    ax2.set_ylabel("volume",color="blue",fontsize=14)
    plt.gca().invert_xaxis()
    plt.title(item_name)        # set filename of figure that will be created
    fig_name = item_name + " Price and Volume.png"
    plt.savefig(fig_name)
    #plt.show(block=False)




    #------------------------------------------------------------------
    # extract and serve analysis_table and data_summary combined info

    # find analysis data on item_name
    analysis = analysis_df.loc[analysis_df['name'] == item_name].values.tolist()

    # extract analysis data on item_name as list (delete item name from list)
    analysis_data = analysis[0]
    analysis_data.reverse()
    analysis_data.pop()
    analysis_data.reverse()

    # extract today's summary data as a list (delete date value from list)
    summary = summary_df[:].values.tolist()
    summary_data = summary[0]
    summary_data.reverse()
    summary_data.pop()
    summary_data.reverse()

    # apply summary data to item_name's analysis data to get expected % increase today
    sum = 0
    calculated_data = []
    changes = []
    change = 0
    for n in range(8):
        change = analysis_data[n] * summary_data[n]
        changes.append(change)
        sum = sum + change
    for n in range(8, 16):
        if analysis_data[n] > 0:
            change = summary_data[n]
        elif analysis_data[n] < 0:
            change = 0 - summary_data[n]
        changes.append(change)
        sum = sum + change

    sum = round(sum, 5)

    # create output dictionary
    output_data = []
    has_effect = []
    fields = ['two_day_run_p', 'three_day_run_p', 'five_day_run_p',
        'seven_day_run_p', 'two_day_run_v', 'three_day_run_v', 'five_day_run_v',
        'seven_day_run_v', 'one_day_avg_p', 'three_day_avg_p', 'seven_day_avg_p', 'fourteen_day_avg_p',
                'one_day_avg_v', 'three_day_avg_v', 'seven_day_avg_v', 'fourteen_day_avg_v']
    for n in range(len(analysis_data)):
        if analysis_data[n] != 0:
            has_effect.append('Yes')
        else:
            has_effect.append('No')
        output_data.append([fields[n], has_effect[n], changes[n]])



        # print all outputs together
    output_df = pd.DataFrame(output_data, columns=['Attribute', 'Has Effect?', 'Effect'])
    print(tabulate(output_df, headers='keys', tablefmt = 'psq1'))
    #plt.show()
    #plt.savefig('graph.png')




    # scrape ge-tracker.com for up-to-date data on item name searched by user
def near_real_data(item_name):
    url = 'https://www.ge-tracker.com/item/'
    head = {'User-Agent':'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/107.0.0.0 Safari/537.36'}
    fields = ['Current Price', 'Current Offer Price', 'Current Sell Price', 'Current Tax',
                   'Buying Quantity (1hr)', 'Selling Quantity (1hr)', 'Buy/Sell Ratio', 'Buy Limit']
    values = {}

        # build url to scrape from item name
    item_name = item_name.lower()
    item_name_for_url = item_name.replace(' ', '-')
    item_name = item_name.capitalize()
    url = url + item_name_for_url

        # go to url
    response = requests.get(url, headers=head)
    old_lines = response.content.splitlines()
    new_lines = []
    ind = 0             # split html lines into list of strings
    for i in range(len(old_lines)):
        new_lines.append(str(old_lines[i]))
    for l in range(len(new_lines)):         # search list of lines line containing item data
        if new_lines[l].__contains__('"name":"%s"' % item_name):
            ind = l

       
            # target line containing item data
    target = new_lines[ind]
    start_ind = target.find('"buyingQuantity"')     # trim unnecessary start/end
    end_ind = target.find('"members"')              # of the line
    target = target[start_ind:end_ind]
    targets_list = target.split(',')
    for t in range(len(targets_list)):
        targets_list[t] = ''.join(char for char in targets_list[t] if char.isdigit())

            # correct the decimal place because "Buy Ratio" sometimes
            # has two or three digits
    if len(targets_list[2]) == 3:
        targets_list[2] = str(int(targets_list[2]) * .01)
    elif len(targets_list[2]) == 2:
        targets_list[2] = str(int(targets_list[2]) * .1)


            # put scraped data values in "values" dictionary
    values[fields[0]] = targets_list[3]
    values[fields[1]] = targets_list[5]
    values[fields[2]] = targets_list[4]
    values[fields[3]] = targets_list[7]
    values[fields[4]] = targets_list[0]
    values[fields[5]] = targets_list[1]
    values[fields[6]] = targets_list[2]
    values[fields[7]] = targets_list[12]

    print(values)




check_files()
analysis_data(item_name)
near_real_data(item_name)